op-node/rollup/driver/sequencer.go

```diff
@@ -159,28 +160,54 @@ func (d *Sequencer) BuildingOnto() eth.L2BlockRef {
 // RunNextSequencerAction starts new block building work, or seals existing work,
 // and is best timed by first awaiting the delay returned by PlanNextSequencerAction.
 // If a new block is successfully sealed, it will be returned for publishing, nil otherwise.
-func (d *Sequencer) RunNextSequencerAction(ctx context.Context) *eth.ExecutionPayload {
+// Only critical errors are bubbled up, other errors are handled internally.
+func (d *Sequencer) RunNextSequencerAction(ctx context.Context) (*eth.ExecutionPayload, error) {
 	if _, buildingID, _ := d.engine.BuildingPayload(); buildingID != (eth.PayloadID{}) {
 		payload, err := d.CompleteBuildingBlock(ctx)
 		if err != nil {
```

Comment:
This function is very branchy, perhaps a switch statement will make it more readable. Its also nearly the same logic at the top and the bottom, perhaps some code could be deduplicated?
- adding time to `nextAction` could be turned into an internal function
- the loglines are the same, maybe a function with a callback where the logs happen internal to the function and the callback executes the actions that differ between the top branches? (`buildingID != (eth.PayloadId{})`?
- perhaps a top level `var` for null payload id to do comparisons against

Just suggestions if you like them

Reply:
Also metrics on any types of errors could be useful here

Reply:
Unfortunately the error checking is using error-targets, not concrete values, so I can't use a switch statement.
And the log-lines are different; the messages indicate which branch we hit.
And I prefer an inline null payload-id check, since it's immutable and a default value.
Good point on metrics, will add those.